core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/event/AbstractEventSupport.java

```diff
@@ -1,19 +1,46 @@
 package org.optaplanner.core.impl.solver.event;
 
+import java.util.Collection;
 import java.util.EventListener;
-import java.util.Set;
-import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.atomic.AtomicLong;
 
 public class AbstractEventSupport<E extends EventListener> {
 
-    protected Set<E> eventListenerSet = new CopyOnWriteArraySet<>();
+    private final AtomicLong ID_COUNTER = new AtomicLong();
+
+    /**
+     * The requirements for this collection are as follows:
+     *
+     * <ul>
+     * <li>Acts on value identity, not equality;</li>
+     * <li>maintains insertion order;</li>
+     * <li>is thread-safe.</li>
+     * </ul>
+     *
+     * In the absence of ConcurrentLinkedIdentityHashSet, this has to do.
+     */
+    private final SortedMap<Long, E> eventListenerMap = new ConcurrentSkipListMap<>();
```

Comment:
Yet another nastiness based on the fact that entity selectors (which end up in this map) can now be equal.

Reply:
I don't understand. EventSupport is per ScoreDirector, no?
Why would it need a Concurrent... if its single threaded?